sigopt/run_params.py

```diff
@@ -0,0 +1,130 @@
+from collections.abc import MutableMapping
+
+from .lib import is_string
+
+
+_get = object.__getattribute__
+_set = object.__setattr__
+
+class RunParameters(MutableMapping):
+  def __init__(self, run_context, fixed_items):
+    _set(self, "__items", dict(fixed_items))
+    _set(self, "__run_context", run_context)
+    _set(self, "__fixed_keys", set(fixed_items.keys()))
+
+  def update(self, *args, **kwds):  # pylint: disable=arguments-differ
+    # this update is atomic, which reduces the number of calls to set_parameter(s)
+    # the default implementation of update would result in a partial update if any of the setters failed
+    # ex. (x := {}).update([(1, 2), ({}, 4)]) => raises TypeError and x == {1: 2}
+    tmp = dict()
+    tmp.update(*args, **kwds)
+    for key in tmp:
+      self.__check_key_type(key)
+    for key in tmp:
+      self.__check_key_is_not_fixed(key)
+    _get(self, "__items").update(tmp)
+    _get(self, "__run_context").set_parameters(tmp)
+
+  def __getattr__(self, attr):
+    try:
+      return self[attr]
+    except KeyError:
+      raise AttributeError(f"no parameter with name {attr!r}")
+
+  def __setattr__(self, attr, value):
+    try:
+      self[attr] = value
+    except KeyError as ke:
+      raise AttributeError(str(ke))
+
+  def __check_key_type(self, key):
+    if not is_string(key):
+      raise TypeError(f"parameter names must be strings, got {type(key).__name__!r}")
+
+  def __check_key_is_not_fixed(self, key):
+    if key in _get(self, "__fixed_keys"):
+      raise ValueError(f"value of {key!r} cannot be changed")
+
+  def __getitem__(self, key):
+    return object.__getattribute__(self, "__items")[key]
```

Comment:
`_get` instead of `object__getattribute__` or is there some nuance?

Reply:
no nuance, I just missed that